File: core/src/main/java/com/mjrfusion/app/allebooks/core/activity/BaseActivity.java
package com.mjrfusion.app.allebooks.core.activity;

import android.os.Bundle;
import android.os.Handler;
import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import com.mjrfusion.app.allebooks.core.R;

public abstract class BaseActivity extends AppCompatActivity {
    public Handler handler;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initializeThisAttributes();
        if (initializeActivity())
            setDefaultToolbar();
        initializeAttributes();
    }

    private void initializeThisAttributes() {
        handler = new Handler(getMainLooper());
    }

    protected abstract boolean initializeActivity();

    private void setDefaultToolbar() {
        setToolbar(R.id.toolbar);
    }

    private void setToolbar(@IdRes int toolbar) {
        setSupportActionBar(findViewById(toolbar));
    }


    public void replaceFragmentBy(
            @IdRes int fragmentContainer,
            @NonNull Class<? extends Fragment> fragment,
            String tag,
            Bundle args,
            boolean addToBackStack) {
        var supportFragmentTransaction = getSupportFragmentManager().beginTransaction()
                .replace(fragmentContainer, fragment, args, tag);
        if (addToBackStack)
            supportFragmentTransaction.addToBackStack(tag);
        supportFragmentTransaction.commit();
    }

    protected abstract void initializeAttributes();
}

File: core/src/main/java/com/mjrfusion/app/allebooks/core/activity/BaseActivity.kt
package com.mjrfusion.app.allebooks.core.activity

import android.os.Bundle
import android.os.Handler
import androidx.annotation.IdRes
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment

abstract class BaseActivity : AppCompatActivity() {
    @JvmField
    var handler: Handler? = null

    @IdRes
    private var defaultToolbarRes = 0

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initializeThisAttributes()
        if (initializeActivity()) setDefaultToolbar()
        initializeAttributes()
    }

    private fun initializeThisAttributes() {
        handler = Handler(mainLooper)
    }

    protected abstract fun initializeActivity(): Boolean

    private fun setDefaultToolbar() {
        setToolbar(defaultToolbarRes)
    }

    private fun setToolbar(@IdRes toolbar: Int) {
        setSupportActionBar(findViewById(toolbar))
    }


    fun replaceFragmentBy(
        @IdRes fragmentContainer: Int,
        fragment: Class<out Fragment?>,
        tag: String? = null,
        args: Bundle? = null,
        addToBackStack: Boolean = false
    ) {
        val supportFragmentTransaction = supportFragmentManager.beginTransaction()
            .replace(fragmentContainer, fragment, args, tag)
        if (addToBackStack) supportFragmentTransaction.addToBackStack(tag)
        supportFragmentTransaction.commit()
    }

    protected abstract fun initializeAttributes()

    fun setDefaultToolbarRes(@IdRes defaultToolbarRes: Int) {
        this.defaultToolbarRes = defaultToolbarRes
    }
}